usaspending_api/download/management/commands/populate_monthly_delta_files.py

```diff
@@ -102,25 +118,111 @@ def create_local_file(self, award_type, source_query, source, agency_code):
-        for split_csv_part in split_csvs:
-            zipped_csvs.write(split_csv_part, os.path.basename(split_csv_part))
+            # Zip the split CSVs into one zipfile
+            zipfile_path = '{}{}_{}_Delta_{}.zip'.format(settings.CSV_LOCAL_PATH, agency_code, award_type,
+                                                         datetime.strftime(date.today(), '%Y%m%d'))
+            logger.info('Creating compressed file: {}'.format(os.path.basename(zipfile_path)))
+            zipped_csvs = zipfile.ZipFile(zipfile_path, 'a', compression=zipfile.ZIP_DEFLATED, allowZip64=True)
+            for split_csv_part in split_csvs:
+                zipped_csvs.write(split_csv_part, os.path.basename(split_csv_part))
+        else:
+            zipfile_path = None
 
         os.close(temp_sql_file)
         os.remove(temp_sql_file_path)
         shutil.rmtree(working_dir)
 
         return zipfile_path
 
+    def add_deletion_records(self, working_dir, award_type, agency_code, source, generate_since):
+        """ Retrieve deletion files from S3 and append necessary records to the end of the the file """
+        logger.info('Retrieving deletion records from S3 files and appending to the CSV')
+        file_path = os.path.join(working_dir, '{}_{}_delta.csv'.
+                                 format(award_type, VALUE_MAPPINGS['transactions']['download_name']))
+
+        # Retrieve all SubtierAgency IDs within this TopTierAgency
+        subtier_agencies = list(SubtierAgency.objects.filter(agency__toptier_agency__cgac_code=agency_code).
+                                values_list('subtier_code', flat=True))
+
+        # Create a list of keys in the bucket that match the date range we want
+        added_rows = False
+        bucket = boto.s3.connect_to_region(settings.BULK_DOWNLOAD_AWS_REGION).get_bucket(settings.FPDS_BUCKET_NAME)
+        for key in bucket.list():
+            match_date = self.check_regex_match(award_type, key.name, generate_since)
+            if match_date:
+                # Create a local copy of the deletion file
+                delete_filepath = '{}{}'.format(working_dir, key.name)
+                key.get_contents_to_filename(delete_filepath)
+                df = pd.read_csv(delete_filepath)
+                os.remove(delete_filepath)
+
+                # Split unique identifier into usable columns and add unused columns
+                df = df[AWARD_MAPPINGS[award_type]['unique_iden']].apply(lambda x: pd.Series(x.split('_'))) \
+                    .replace('-none-', '', regex=True).replace('-NONE-', '', regex=True) \
+                    .rename(columns=AWARD_MAPPINGS[award_type]['column_headers'])
+
+                # Only include records within the correct agency, and populated files
+                if len(df.index) == 0:
+                    continue
+                if agency_code != 'all':
+                    df = df[df['awarding_sub_agency_code'].isin(subtier_agencies)]
```

Comment:
Clarification: we're using `subtier_code`s to check against toptier agencies as toptier agencies could be duplicates?

Reply:
The deletion records do not have any information about top tier agencies, but does have sub tier agency information. So what we do here is make sure that the `awarding_sub_agency_code` maps to the top tier agency.